### PR TITLE
Catalin/pcchair 10/query jobs

### PR DESCRIPTION
**Pull request by cata-b**
- Added a flexible querying endpoint for getting jobs
- Tested the endpoint
- Installed new packages: FluentValidation and Moq

### CODE REVIEW

File: project-backend.Test/ControllerTests/JobControllerTest.cs [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+            };
+
+            _dbContext.Users.AddRange(user1, user2);
+            _dbContext.SaveChanges();
+
+            _dbContext.Jobs.AddRange(
+                    new JobDAO
+                    {
+                        Name = "Job1",
+                        Description = "Job1Desc",
```
Comment (AndrewBarzu) [inline project-backend.Test/ControllerTests/JobControllerTest.cs]: For tests in general you can use the arrange/act/assert setup. It's usual for backend testing to be done this way. It's not necessary but it just formalizez how tests look. Also, you are testing the controller here, but you also use the provider, so you're basically testing that as well, making it more of an integration test than an unit test. You should instead just mock the provider to return what it has to, without having to create a database context and the sort.

**Review by AndreiPetrusMihai — APPROVED** [review]

File: project-backend/Controllers/JobController.cs [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+            var jobs = _jobProvider.QueryJobs();
+            if (query.ByCurrentUserOnly ?? false)
+            {
+                jobs = jobs.Where(j => j.UserId == userId);
+            }
+
+            if (query.OlderThan != null)
+            {
+                jobs = jobs.Where(j => j.PostDate <= query.OlderThan);
```
Comment (AndrewBarzu) [inline project-backend/Controllers/JobController.cs]: This method is quite large and should be split in some smaller methods

File: project-backend/Models/Validators/JobController/GetJobs/GetJobsQueryValidator.cs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+            RuleFor(x => x.FilterFields).NotEmpty().Unless(x => x.ExactFilters == null && x.FilterValues == null).WithMessage(filtersErrorMessage);
+
+            // the lengths of the filters are equal
+            RuleFor(x => x.FilterValues.Length)
+                .Equal(x => x.FilterFields.Length).WithMessage(filtersErrorMessage)
+                .Equal(x => x.ExactFilters.Length).WithMessage(filtersErrorMessage)
+                .When(x => x.FilterValues != null && x.FilterFields != null && x.ExactFilters != null);
+
+            // similar checks for OrderBy and Ascending
+            RuleFor(x => x.OrderBy).NotEmpty().Unless(x => x.Ascending == null).WithMessage(orderByErrorMessage);
```
Comment (AndrewBarzu) [inline project-backend/Models/Validators/JobController/GetJobs/GetJobsQueryValidator.cs:26]: This is really nice. Didn't even know this was possible :)

File: project-backend/Controllers/JobController.cs (hunk 2 above)
Comment (cata-b) [reply project-backend/Controllers/JobController.cs]: Fixed this in https://github.com/PC-Chair-Fight/project-backend/pull/17/commits/9bdf23967d4bc518ac157d157f3942ac417c833f. Please review the changes again.

File: project-backend.Test/ControllerTests/JobControllerTest.cs (hunk 1 above)
Comment (cata-b) [reply project-backend.Test/ControllerTests/JobControllerTest.cs]: Fixed in https://github.com/PC-Chair-Fight/project-backend/pull/17/commits/8e24bcc8257ebf4c10069b01a9af7a3efb14f3ee

**Review by AndrewBarzu — APPROVED** [review]

**Pull request merged**